### PR TITLE
Enable Samsung built-in resume functionality

### LINKED ISSUE #665: Support for Samsung TV C5000's built-in Resume function
The TV's built-in media player has the function of resuming a previously stopped video using the remote's "D" button. I know that UMS already has a similar functionality but it transcodes the video on the fly and there are some limitations to it - such as you cannot FF or rewind and it's not as precise as the built-in option.
It seems like the same issue is already discussed here: http://forum.serviio.org/viewtopic.php?p=6504&sid=308fb26b61282ae7fa65fcd191ad4e7a#p6504
I can supply you with a wireshark dump if necessary. Thank you.

### PR DESCRIPTION
**Pull request by drakulis**
Solution taken from: http://www.serviio.org/forum/viewtopic.php?f=3&t=272&start=0#p6504

Closes #665

### CODE REVIEW

**Comment by SubJunk** [comment]
@drakulis this is a very exciting addition, thanks! I will review it tonight

File: src/main/java/net/pms/network/RequestV2.java [hunk 1]
```diff
@@ -895,6 +911,8 @@ public ChannelFuture answer(
 				response.append(CRLF);
 				response.append(HTTPXMLHelper.SOAP_ENCODING_FOOTER);
 				response.append(CRLF);
+			} else {
+				LOGGER.info("Unsupported action received: " + content);
```
Comment (SubJunk) [inline src/main/java/net/pms/network/RequestV2.java]:
```suggestion
				LOGGER.debug("Unsupported action received: " + content);
```

File: src/main/java/net/pms/network/RequestV2.java [hunk 2]
```diff
@@ -31,12 +33,18 @@
 import java.util.Map.Entry;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+
```
Comment (SubJunk) [inline src/main/java/net/pms/network/RequestV2.java]:
```suggestion
```

File: src/main/java/net/pms/database/TableFilesStatus.java [hunk 3]
```diff
@@ -237,6 +237,76 @@ public static Boolean isFullyPlayed(final String fullPathToFile) {
...
+				try (ResultSet resultSet = statement.executeQuery(query)) {
+					if (resultSet.next()) {
+						result = resultSet.getInt("BOOKMARK");
+					}
+				}
+			} finally {
+				TABLE_LOCK.readLock().unlock();
+			}
+		} catch (SQLException e) {
+			LOGGER.error("Database error while looking up file status in " + TABLE_NAME + " for \"{}\": {}", fullPathToFile, e.getMessage());
```
Comment (SubJunk) [inline src/main/java/net/pms/database/TableFilesStatus.java]:
```suggestion
			LOGGER.error("Database error while looking up file bookmark in " + TABLE_NAME + " for \"{}\": {}", fullPathToFile, e.getMessage());
```

**Comment by SubJunk** [comment]
@drakulis also just a note, I don't have a Samsung device to test on, so I am trusting you that this works :)

File: src/main/java/net/pms/database/TableFilesStatus.java [hunk 4]
```diff
@@ -371,6 +441,12 @@ private static void upgradeTable(final Connection connection, final int currentV
 
 						version = 8;
 						break;
+					case 8:
+						try (Statement statement = connection.createStatement()) {
+							statement.execute("ALTER TABLE " + TABLE_NAME + " ADD BOOKMARK INTEGER DEFAULT 0");
```
[the remark this answers is not in the document]
Comment (drakulis) [reply src/main/java/net/pms/database/TableFilesStatus.java]: Ups, overlooked. Done.

File: src/main/java/net/pms/database/TableFilesStatus.java (hunk 3 above)
Comment (drakulis) [reply src/main/java/net/pms/database/TableFilesStatus.java]: done

File: src/main/java/net/pms/network/RequestV2.java (hunk 2 above)
Comment (drakulis) [reply src/main/java/net/pms/network/RequestV2.java]: removed empty line.

File: src/main/java/net/pms/network/RequestV2.java (hunk 1 above)
Comment (drakulis) [reply src/main/java/net/pms/network/RequestV2.java]: done

**Comment by drakulis** [comment]
@SubJunk I've just pushed changes regarding your comments.

**Comment by drakulis** [comment]
I've implemented it using my Samsung TV (D-serie). Here is some comparison how it was before and after my change
ums-samsung-resume

**Comment by SubJunk** [comment]
@drakulis thanks for your submission! We will have a lot of happy users I'm sure.
I wonder if there is a related optimisation we can do where we hide the "Resume" files that UMS creates if the device is Samsung, since they are redundant now. Do you think that would be good?

**Comment by drakulis** [comment]
Well... I would leave this option. Its an option so could be disabled if user doesn't want it.
This option may be usable eg. on Samsung Android devices like tablets where some third party app is used. UMS will recognize such device as a Samsung device (which in fact it is) but DLNA app may not support Samsung bookmarking. What I wanted to say is already described in following issue #1639

**Comment by SubJunk** [comment]
Good point. Another optimization of this work would be to make this a renderer setting instead of using `isSamsung()`

**Pull request merged**